### ISSUE TITLE
erts_code_purger hangs on `cpc_receive` during purge

### ISSUE THREAD

**Issue by josevalim** [erlang/otp#8119]
**Describe the bug**

Hi folks, unfortunately we were not able to isolate this bug, but I am posting it here to see if you have any ideas.

We are finding that the `erts_code_purger` process gets stuck forever on `cpc_receive`. The Elixir compiler uses temporary modules and purges (proper `purge`, not `soft_purge`) them as code compilation happens. Sometimes, however, compilation hangs due to purging. Removing the calls to `purge` completely solves the compilation issues, which helped us pinpoint the problem.

Furthermore, as we attempted to debug this, simple changes to `erts_code_purger` fixed the problem. In particular, if we add the following:

```
after 1_000 -> cpc_receive(CpcS, ReqInfo, KillState0, PReqs)
```

to:

https://github.com/erlang/otp/blob/aa94ab2cdb5b21882bfc50c64fb1b1dab18ef379/erts/preloaded/src/erts_code_purger.erl#L285

The bug is fixed, which makes me think this is a not a logic bug in `erts_code_purger`, but rather some race condition on the runtime?

I also found we had a similar report in the past, discussed in #4331. I am looking into making crash dumps available, but they were not useful in #4331 and I am afraid they may not be useful here. :( 

**To Reproduce**

Unfortunately no straight-forward mechanism for now.

**Affected versions**

OTP 26, maint. We didn't test earlier versions.

**Comment by rickard-green** [comment]
A `core` dump (`kill -ABRT <BEAM-PID>`) would be more helpful than a crash dump when this happens. Also the `beam.smp` executable used is needed if you are able to produce a `core` when this has happens.

**Comment by rickard-green** [comment]
@josevalim Please try with #8201 merged into latest maint (which contains 8038d202c4785385b273eff6e10697c1c1f73b33 which fixes a potential hang of `code_server`)

**Comment by amalbuquerque** [comment]
@rickard-green with the latest `maint` + cherry picking #8201 single commit:

```
commit 686c8a4a7fe4fa92ee219fb5419b7671d6c8ef49
Author: Rickard Green <rickard@erlang.org>
Date:   Wed Feb 28 12:42:24 2024 +0100

    [erts] Fix signal wakeup issue

diff --git a/erts/emulator/beam/erl_proc_sig_queue.c b/erts/emulator/beam/erl_proc_sig_queue.c
index 87decd2642..e048de0b45 100644
--- a/erts/emulator/beam/erl_proc_sig_queue.c
+++ b/erts/emulator/beam/erl_proc_sig_queue.c
@@ -9128,7 +9128,7 @@ erts_proc_sig_queue_try_enqueue_to_buffer(Eterm from,
          * we have new signals.
          */

-        if (len && !nonempty_slots_before) {
+        if (!nonempty_slots_before) {

             /*
              * There is one situation in which we need to synchronize

commit b6c6537f45458b326a867a29ceda6ecdf46abc39 (origin/maint)
Merge: dc2024d3b7 5da1a5575f
Author: Lukas Larsson <lukas@erlang.org>
Date:   Thu Feb 29 13:03:37 2024 +0100

    Merge branch 'lukas/kernel/fix-io-get_password/OTP-19007' into maint

    * lukas/kernel/fix-io-get_password/OTP-19007:
      kernel: Fix io:get_password
```

I still see the compilation freeze.

**Comment by rickard-green** [comment]
Thanks! Found another bug, will soon push a new fix.

**Comment by rickard-green** [comment]
Please try the new fix in #8201

**Comment by amalbuquerque** [comment]
> Please try the new fix in #8201

Great news @rickard-green ! With the #8201 merged into the latest `main` (b6c6537f45458b326a867a29ceda6ecdf46abc39) I was able to successfully compile 10 times in a row without the compilation freeze happening :tada:

**Comment by rickard-green** [comment]
Great! Thanks!

**Comment by rickard-green** [comment]
The fix has been released in the [OTP 26.2.3](https://erlang.org/download/OTP-26.2.3.README) and [OTP 25.3.2.10](https://erlang.org/download/OTP-25.3.2.10.README) patches.
